### DISCUSSION TITLE
Stuck when migrating to git-lfs when repo has 10,000s of refs

### DISCUSSION THREAD

**Discussion by thejewdude** [git-lfs/git-lfs#4351] [Q&A]
I'm in the process of migrating a repo to a new hosting provider (its GitHub!) and before migrating the repo to git-lfs I used git-filter-repo to remove tons of large files that we no longer need and didn't want stored in git-lfs at all. The tool created about 50,000 replace refs.

When running `git lfs migrate import --everything` the process really slows down when it updates each replace ref to point to the new commit. It seemed to move pretty quickly through the first 1000 or so refs but after that it started to fall off a cliff. I let the command run for 24 hours but with no end in sight I pulled the plug. After deleting all the replace refs, I tried again and the import process completed in reasonable amount of time. 

Is there any way I can make the process complete quicker with so many refs? And is it actually a bug that at first the update-refs complete quickly but then really slow down?

(Happy to provide any more details)

**Comment by bk2204** [comment]
I think part of the problem here is that when you replace the entire history with `git lfs migrate import`, you're walking each commit at most once, since by the time we get to a given commit, we'll have rewritten each object in the preceding history.  When you have the replace refs, you also have to walk each object in the replace ref and rewrite it, which may be drastically different and may point to an entirely different history.  To the extent that Git LFS operates on them, it does so using Git, and they are not especially performant in Git.

So my recommendation here would be that you remove the replace refs before migrating to Git LFS, since Git LFS doesn't necessarily honor them and once you rewrite the history, they'll be useless anyway.

**Reply by thejewdude** [reply]
Thanks for responding!
I forgot to mention that I'm running import with the `--everything` option, but I'm not sure understanding you. Shouldn't Git LFS treat replace refs the same way it does other refs? I'm assuming this means that Git LFS knows not to update commits that it already updated from a different ref. I'm also assuming that git-lfs should update the commit that the replace refs are pointing to in the same way it can update the commit that a regular ref is pointing to?

Removing the replace refs before migrating did indeed help. But I'm hoping I can keep the replace refs in case that someone needs to refer back to old commit ids (say from old pull requests). 

Has it been discussed to have git-lfs output replace refs to make migrations less painless?
